File: app/src/main/java/org/bspb/smartbirds/pro/ui/fragment/MonitoringEntryListFragment.java
package org.bspb.smartbirds.pro.ui.fragment;

import android.app.AlertDialog;
import android.app.ListFragment;
import android.content.Context;
import android.content.DialogInterface;
import android.database.Cursor;
import android.os.Bundle;
import android.util.Log;
import android.view.ActionMode;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.CursorAdapter;
import android.widget.ListView;

import org.androidannotations.annotations.AfterInject;
import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.Bean;
import org.androidannotations.annotations.EFragment;
import org.androidannotations.annotations.FragmentArg;
import org.bspb.smartbirds.pro.R;
import org.bspb.smartbirds.pro.SmartBirdsApplication;
import org.bspb.smartbirds.pro.adapter.ModelCursorAdapter;
import org.bspb.smartbirds.pro.adapter.ModelCursorFactory;
import org.bspb.smartbirds.pro.beans.MonitoringCursorEntries;
import org.bspb.smartbirds.pro.content.MonitoringEntry;
import org.bspb.smartbirds.pro.content.MonitoringManager;
import org.bspb.smartbirds.pro.enums.EntryType;
import org.bspb.smartbirds.pro.service.DataOpsService_;
import org.bspb.smartbirds.pro.ui.partial.MonitoringEntryListRowPartialView;
import org.bspb.smartbirds.pro.ui.partial.MonitoringEntryListRowPartialView_;

import java.util.Locale;


/**
 * Created by groupsky on 08.03.17.
 */

@EFragment
public class MonitoringEntryListFragment extends ListFragment implements MonitoringCursorEntries.Listener {

    private static final String TAG = SmartBirdsApplication.TAG + ".FFormLst";

    private CursorAdapter adapter;

    @Bean
    MonitoringCursorEntries entries;

    @Bean
    MonitoringManager monitoringManager;

    private String monitoringCode;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        Log.d(TAG, "onCreate");
        super.onCreate(savedInstanceState);

        adapter = new ModelCursorAdapter<MonitoringEntry>(getActivity(), R.layout.partial_monitoring_entry_list_row, entries != null ? entries.getCursor() : null, new ModelCursorFactory<MonitoringEntry>() {
            @Override
            public MonitoringEntry createModelFromCursor(Cursor cursor) {
                return MonitoringManager.entryFromCursor(cursor);
            }
        }) {
            @Override
            public View newView(Context context, Cursor cursor, ViewGroup parent) {
                return MonitoringEntryListRowPartialView_.build(context);
            }

            @Override
            public void bindView(View view, Context context, MonitoringEntry model) {
                if (!(view instanceof MonitoringEntryListRowPartialView))
                    throw new IllegalArgumentException("Must use " + MonitoringEntryListRowPartialView.class.getSimpleName());
                ((MonitoringEntryListRowPartialView) view).bind(model);
            }
        };
        setListAdapter(adapter);
    }

    @AfterInject
    protected void setupLoader() {
        entries.setMonitoringCode(monitoringCode);
        entries.setListener(this);
        if (adapter != null) adapter.swapCursor(entries.getCursor());
    }

    @AfterViews
    protected void setupListview() {
        final ListView lv = getListView();
        setEmptyText(getText(R.string.emptyList));
        lv.setChoiceMode(AbsListView.CHOICE_MODE_MULTIPLE_MODAL);
        lv.setMultiChoiceModeListener(new AbsListView.MultiChoiceModeListener() {
            @Override
            public void onItemCheckedStateChanged(ActionMode mode, int position, long id, boolean checked) {
                Log.d(TAG, String.format(Locale.ENGLISH, "onItemCheckedStateChanged: %d %s", position, checked));
            }

            @Override
            public boolean onCreateActionMode(ActionMode mode, Menu menu) {
                Log.d(TAG, String.format(Locale.ENGLISH, "onCreateActionMode: %s", mode));
                getActivity().getMenuInflater().inflate(R.menu.listselection, menu);
                return true;
            }

            @Override
            public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
                return false;
            }

            @Override
            public boolean onActionItemClicked(final ActionMode mode, MenuItem item) {
                Log.d(TAG, String.format(Locale.ENGLISH, "onActionItemClicked: %s", item));
                switch (item.getItemId()) {
                    case R.id.action_select_all:
                        for (int i = 0; i < lv.getCount(); i++)
                            lv.setItemChecked(i, true);
                        return true;
                    case R.id.action_delete:
                        final long[] selectedItems = lv.getCheckedItemIds();

                        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());

                        builder.setMessage(getString(R.string.confirm_delete_n, selectedItems.length));
                        builder.setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                mode.finish();
                                monitoringManager.deleteEntries(selectedItems);
                                DataOpsService_.intent(getActivity()).generateMonitoringFiles(monitoringCode).start();
                            }
                        });
                        builder.setNegativeButton(android.R.string.cancel, null);

                        AlertDialog dialog = builder.create();
                        dialog.show();
                        return true;
                }
                return false;
            }

            @Override
            public void onDestroyActionMode(ActionMode mode) {
                Log.d(TAG, String.format(Locale.ENGLISH, "onDestroyActionMode: %s", mode));
            }
        });
    }

    @Override
    public void onListItemClick(ListView l, View v, int position, long id) {
        super.onListItemClick(l, v, position, id);
        MonitoringEntry entry = (MonitoringEntry) getListAdapter().getItem(position);
        if (entry != null && getActivity() instanceof Listener) {
            ((Listener) getActivity()).onMonitoringEntrySelected(entry.id, entry.type);
        }
    }

    @FragmentArg
    public void setMonitoringCode(String monitoringCode) {
        this.monitoringCode = monitoringCode;
        if (entries != null) entries.setMonitoringCode(monitoringCode);
    }

    @Override
    public void onMonitoringEntriesChanged(MonitoringCursorEntries entries) {
        if (adapter != null) adapter.swapCursor(entries.getCursor());
    }

    public interface Listener {
        void onMonitoringEntrySelected(long id, EntryType entryType);
    }
}

File: app/src/main/kotlin/org/bspb/smartbirds/pro/ui/fragment/MonitoringEntryListFragment.kt
package org.bspb.smartbirds.pro.ui.fragment

import android.app.AlertDialog
import android.os.Bundle
import android.util.Log
import android.view.ActionMode
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AbsListView
import android.widget.AbsListView.MultiChoiceModeListener
import android.widget.ListView
import androidx.fragment.app.ListFragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.launch
import org.bspb.smartbirds.pro.R
import org.bspb.smartbirds.pro.SmartBirdsApplication
import org.bspb.smartbirds.pro.adapter.MonitoringEntryListAdapter
import org.bspb.smartbirds.pro.content.Monitoring
import org.bspb.smartbirds.pro.content.MonitoringEntry
import org.bspb.smartbirds.pro.enums.EntryType
import org.bspb.smartbirds.pro.utils.MonitoringManager
import org.bspb.smartbirds.pro.viewmodel.MonitoringEntryListViewModel
import java.util.Locale

open class MonitoringEntryListFragment : ListFragment() {

    companion object {
        private const val TAG = SmartBirdsApplication.TAG + ".FFormLst"

        const val ARG_MONITORING_CODE = "monitoringCode"

        fun newInstance(monitoringCode: String?): MonitoringEntryListFragment {
            return MonitoringEntryListFragment().apply {
                arguments = Bundle().apply {
                    monitoringCode?.let {
                        putString(ARG_MONITORING_CODE, monitoringCode)
                    }
                }
            }
        }
    }

    private var adapter: MonitoringEntryListAdapter? = null

    protected val monitoringManager = MonitoringManager.getInstance()

    protected var code: String? = null
    protected var monitoring: Monitoring? = null

    private val viewModel: MonitoringEntryListViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        Log.d(TAG, "onCreate")
        readArgs()
        setupLoader()
        super.onCreate(savedInstanceState)

        adapter = MonitoringEntryListAdapter(requireContext())
        listAdapter = adapter
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupListview()
    }

    private fun readArgs() {
        setMonitoringCode(arguments?.getString(ARG_MONITORING_CODE))
    }

    private fun initViewModel() {
        viewModel.init(code)
        viewModel.entries?.observe(viewLifecycleOwner) {
            adapter?.apply {
                clear()
                addAll(it)
                notifyDataSetChanged()
            }
        }
    }

    private fun setupLoader() {
        code?.let {
            lifecycleScope.launch {
                monitoring = monitoringManager.getMonitoring(it)
            }
        }
    }

    private fun setupListview() {
        initViewModel()
        val lv = listView
        setEmptyText(getText(R.string.emptyList))
        if (monitoring?.status == Monitoring.Status.uploaded) {
            return
        }

        lv.choiceMode = AbsListView.CHOICE_MODE_MULTIPLE_MODAL
        lv.setMultiChoiceModeListener(object : MultiChoiceModeListener {
            override fun onItemCheckedStateChanged(
                mode: ActionMode,
                position: Int,
                id: Long,
                checked: Boolean
            ) {
                Log.d(
                    TAG,
                    String.format(
                        Locale.ENGLISH,
                        "onItemCheckedStateChanged: %d %s",
                        position,
                        checked
                    )
                )
            }

            override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
                Log.d(TAG, String.format(Locale.ENGLISH, "onCreateActionMode: %s", mode))
                activity?.menuInflater?.inflate(R.menu.listselection, menu)
                return true
            }

            override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
                return false
            }

            override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
                Log.d(TAG, String.format(Locale.ENGLISH, "onActionItemClicked: %s", item))
                when (item.itemId) {
                    R.id.action_select_all -> {
                        var i = 0
                        while (i < lv.count) {
                            lv.setItemChecked(i, true)
                            i++
                        }
                        return true
                    }

                    R.id.action_delete -> {
                        val selectedItems = lv.checkedItemIds
                        val builder = AlertDialog.Builder(activity)
                        builder.setMessage(getString(R.string.confirm_delete_n, selectedItems.size))
                        builder.setPositiveButton(android.R.string.ok) { _, _ ->
                            lifecycleScope.launch {
                                mode.finish()
                                monitoringManager.deleteEntries(selectedItems)
                            }
                        }
                        builder.setNegativeButton(android.R.string.cancel, null)
                        val dialog = builder.create()
                        dialog.show()
                        return true
                    }
                }
                return false
            }

            override fun onDestroyActionMode(mode: ActionMode) {
                Log.d(TAG, String.format(Locale.ENGLISH, "onDestroyActionMode: %s", mode))
            }
        })
    }

    override fun onListItemClick(l: ListView, v: View, position: Int, id: Long) {
        super.onListItemClick(l, v, position, id)
        val entry = listAdapter?.getItem(position) as MonitoringEntry
        if (activity is Listener) {
            (activity as Listener).onMonitoringEntrySelected(entry.id, entry.type)
        }
    }

    fun setMonitoringCode(monitoringCode: String?) {
        this.code = monitoringCode
        this.code?.let {
            lifecycleScope.launch {
                monitoring = monitoringManager.getMonitoring(it)
            }
        }

    }

    interface Listener {
        fun onMonitoringEntrySelected(id: Long, entryType: EntryType?)
    }
}